Improve AppView loading behavior

andresmgot opened the pull request:
Fixes #748
Requires #847

The goal of #748 is to provide feedback to the user when the URLs table is loading. In this PR I have added the LoadingWrapper to all the tables in the AppView page (not only to the URLs one):

Summary of the changes:

 - The URLs table now have a title: "Access URLs"
 - All the tables shows the loading icon if the content is not ready.
 - In case there are no elements for a table a message like "No resources found" is shown.
 - I have removed the component AppDetails to directly use `DeploymentTable`, `ServiceTable` and `OtherResourcesTable`.
 - I have removed the title "Details" since it was splitting the Secrets and the rest of resources.

Todo:

 - [x] Unit tests to check that each table behaves as a loading component and to check the 0 resources scenario.

![gif_](https://user-images.githubusercontent.com/4025665/49296053-87d6e480-f4b7-11e8-9a7c-521a1b389545.gif)

andresmgot commented:
Sorry, I forgot to change the base branch. Now it's compared with the `secretButton` branch.

andresmgot commented:
Thanks for the comments @prydonius. I will stop working on this until #847 is completed to avoid pointing to a moving target.

migmartri commented on dashboard/src/components/AppView/AccessURLTable/AccessURLTable.tsx:
```diff
@@ -6,44 +6,48 @@ import { GetURLItemFromIngress } from "./AccessURLItem/AccessURLIngressHelper";
 interface IServiceTableProps {
-  services: { [s: string]: IResource };
-  ingresses: { [i: string]: IResource };
+  services: IResource[];
```
Could we instead of using IResource use something more specific in reference to the Service? I am not sure if we have the types for that though.

andresmgot replied:
we don't have a type `IService` but it's not needed since it follows the generic layout (metadata, spec and status keys).

migmartri commented on dashboard/src/components/AppView/AppView.test.tsx:
```diff
@@ -259,25 +253,42 @@ describe("AppViewComponent", () => {
+      wrapper.setState({ ingresses });
+
+      expect(wrapper.find(AccessURLTable)).not.toExist();
+      const loading = wrapper.find(LoadingWrapper);
```
Maybe you can reuse `itBehavesLike(aLoadingComponent)`

https://github.com/kubeapps/kubeapps/blob/33d6bf4a47c3e4a3e95ab9a59d55cdb6bf8371fb/dashboard/src/components/UpgradeForm/UpgradeForm.test.tsx#L14

andresmgot replied:
`itBehavesLike("aLoadingComponent", ` checks if there is a LoadingWrapper within the given component. We cannot use that here because:

 - If we pass `AppView` as the component it may find LoadingWrapper not related to the component we are interested in. For example it can detect a LoadingWrapper around the ServiceTable but we are interested in the AccessURLTable.
 - If we pass the component that we want (AccessURLTable for example) it won't work because the LoadingWrapper is actually outside the component.

andresmgot replied:
After some refactoring I have moved the different LoadingWrappers so I can use now the `itBehavesLike` method.

migmartri commented on dashboard/src/components/AppView/AccessURLTable/AccessURLTable.tsx:
```diff
@@ -6,44 +6,48 @@ import { GetURLItemFromIngress } from "./AccessURLItem/AccessURLIngressHelper";
 
 interface IServiceTableProps {
-  services: { [s: string]: IResource };
-  ingresses: { [i: string]: IResource };
```
Why has the format changed? Is it consistent across the other tables?

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -126,22 +126,26 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
+      const id = `${i.kind}/${i.metadata.name}`;
       switch (i.kind) {
         case "Deployment":
+          this.setState({ deployments: { ...this.state.deployments, [id]: { isFetching: true } } });
```
Why don't you initialize the state with isFetching: true and remove these lines? 

Also, does this line ever set deployments? It seems that the isFetching thing is what is relevant isn't it?

andresmgot replied:
> Why don't you initialize the state with isFetching: true and remove these lines?

I am not sure if I understand this comment. Note that that the `isFetching` property depends on the independent elements of the array, I cannot initialize `isFetching` property.

> Also, does this line ever set deployments? It seems that the isFetching thing is what is relevant isn't it?

True, I am only interested in the isFetching property

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -160,20 +164,26 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
   }
 
-  public get isLoading(): boolean {
+  public get isAppLoading(): boolean {
```
You might want to explain in a comment what does loading app really means.

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -217,23 +233,32 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
                   <div className="col-4">
-                    <DeploymentStatus deployments={this.deploymentArray()} info={app.info!} />
+                    <DeploymentStatus
+                      deployments={this.arrayFromState("deployments")}
```
you could use `deployments` defined above no? 

```
const deployments = this.arrayFromState("deployments");
```

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -266,56 +291,77 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
+    return res;
+  }
+
+  // Retrieve is some deployments/service/ingresses is still loading
```
> ingresses is still loading

ingresses are still loading

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -37,13 +39,12 @@ export interface IAppViewProps {
 interface IAppViewState {
-  deployments: { [d: string]: IResource };
+  manifest: IResource[];
+  deployments: { [d: string]: { resource?: IResource; isFetching: boolean } };
```
shouldn't isFetching be outside of the struct? I mean, loading means loading all the resources of that type but not loading a single one isn't it?

Or we are planning to load resources individually?

migmartri replied:
My bad, I think that we actually fetch resources individually, although personally I am not sure if it makes sense to have this flag per resource in our case.

andresmgot replied:
yes, it's per element right now.

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -266,56 +291,77 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
+
+  // Retrieve is some deployments/service/ingresses is still loading
+  private isLoading(type: string) {
+    return _.some(this.state[type], e => e.isFetching);
```
See my comment about having isFetching at the resources level above.

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -266,56 +291,77 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
     return error;
   };
 
+  private findLoadingResource = (resources: { [s: string]: IKubeItem }): boolean => {
```
Could you explain why is this function needed?

andresmgot replied:
it's convenient to find a loading item from a list. I've moved this to a helper function.

migmartri commented on dashboard/src/components/AppView/OtherResourcesTable.tsx:
```diff
@@ -0,0 +1,36 @@
+import { IResource } from "../../shared/types";
+
+interface IAppDetailsProps {
+  otherResources: { [r: string]: IResource };
```
Shouldn't this for consistency be `IResource[];` ?

migmartri commented on dashboard/src/components/AppView/DeploymentTable.tsx:
```diff
@@ -4,29 +4,33 @@ import { IResource } from "../../shared/types";
 
 interface IDeploymentTableProps {
-  deployments: { [d: string]: IResource };
+  deployments: IResource[];
```
Could this be Ideployment[]?

andresmgot replied:
with deployments happen the same than services, since it follows the same layout it's not necessary to define a new interface.

migmartri replied:
Sorry, I think that I do not understand this comment.

andresmgot commented:
I will wait for #847 to be finished before keep working on this.

andresmgot commented:
@prydonius @migmartri this should be ready to be reviewed again. Since the last review I have:

 - Moved each table to a folder
 - Moved the loading check to the table component (not in the AppView)
 - Standardized how deployments/secrets/ingresses are represented (using `Array<IKubeItem<IResource>>`
 - Wait for the deployments to load before rendering the `DeploymentStatus` box

andresmgot commented:
@prydonius @migmartri this should be ready to be reviewed again. Since the last review I have:

 - Moved each table to a folder
 - Moved the loading check to the table component (not in the AppView)
 - Standardized how deployments/secrets/ingresses are represented (using `Array<IKubeItem<IResource>>`
 - Wait for the deployments to load before rendering the `DeploymentStatus` box

prydonius commented on dashboard/src/components/AppView/AccessURLTable/AccessURLTable.tsx:
```diff
@@ -1,49 +1,76 @@
   public render() {
-    const { services, ingresses } = this.props;
+    const { ingresses, services } = this.props;
+    let ingressSection = <p>The current application does not expose a public URL.</p>;
```
whilst both services and ingresses are ingresses, the term is confusing because at first I thought we were excluding services here. maybe we should call this accessTable or something else more generic?

prydonius commented on dashboard/src/components/AppView/AccessURLTable/AccessURLTable.tsx:
```diff
@@ -1,49 +1,76 @@
 interface IServiceTableProps {
-  services: { [s: string]: IResource };
-  ingresses: { [i: string]: IResource };
+  services: Array<IKubeItem<IResource>>;
```
if we have IService and IIngress, could this be `Array<IKubeItem<IService | IIngress>>`? would help us in not having to do the type conversion below.

prydonius commented on dashboard/src/components/AppView/AppView.test.tsx:
```diff
@@ -241,22 +226,26 @@ describe("AppViewComponent", () => {
-      } as IResource;
+            ],
+          } as IIngressSpec,
+        } as IResource,
```
this is quite messy, ideally we should just be able to use IIngress here and not have to convert it twice 
(once for the spec, and for the full thing)

prydonius commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -113,29 +115,38 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
 
     const sockets: WebSocket[] = [];
     let secrets = {};
+    let deployments = {};
```
what are these for? aren't we already setting them in the state in `handleEvent`?

migmartri replied:
I am assuming that it is to set an initial state based on the value in the manifest?

If that's the case, I'd add a comment.

prydonius commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -147,23 +158,32 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
 
-  public get isLoading(): boolean {
+  // isAppLoading checks if the given app has been collected from tiller
+  public get isAppLoading(): boolean {
```
why do we need a getter instead of a normal function? I'm just curious mostly, but I also think explicit functions are better because I now can't tell if this is a property or a function when I see it being used. Why couldn't appInfo be a getter?

andresmgot replied:
this was a getter before and I didn't change it. I don't have a good reason why this is not a function. Actually I don't think this function is useful anyway.

prydonius commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -259,8 +274,17 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
-  private deploymentArray(): IResource[] {
-    return Object.keys(this.state.deployments).map(k => this.state.deployments[k]);
+  // Retrieve the deployments/service/ingresses if they are already loaded
+  private arrayFromState(type: string): Array<IKubeItem<any>> {
```
isn't it going to be an `IKubeItem<IDeployment | IService | IIngress>` or if we need to be more general `IKubeItem<IResource>`. We should try to avoid using `any` if possible.

andresmgot replied:
_removed this method_

prydonius commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -259,8 +274,17 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
+  private arrayFromState(type: string): Array<IKubeItem<any>> {
+    const elems = Object.keys(this.state[type]);
+    const res: Array<IKubeItem<any>> = [];
+    elems.forEach(e => {
```
couldn't this use `Array.filter` instead of populating a new array?

prydonius commented on dashboard/src/components/AppView/helpers/index.ts:
```diff
@@ -0,0 +1,13 @@
+
+export default function isSomeResourceLoading(resources: Array<IKubeItem<any>>): boolean {
+  let isFetching = false;
+  _.each(resources, i => {
```
couldn't this be simplified with an `Array.any` or `_.any`?

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -29,11 +31,11 @@ export interface IAppViewProps {
+  deployments: { [d: string]: IKubeItem<IResource> };
+  services: { [s: string]: IKubeItem<IResource> };
+  ingresses: { [i: string]: IKubeItem<IResource> };
   secrets: { [s: string]: ISecret };
```
Why some are IkubeItem and others are Iresrouce/ISecret?

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -147,23 +158,32 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
     const resource: IResource = msg.object;
-    const key = `${resource.kind}/${resource.metadata.name}`;
+    const item = {
+      [resource.metadata.name]: { item: resource, isFetching: false },
```
So the key is just the resource name and not the `resourceType/resourceName` ? In that case we will not be able to flatten resources in the future no?

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -29,11 +31,11 @@ export interface IAppViewProps {
-  otherResources: { [r: string]: IResource };
-  services: { [s: string]: IResource };
-  ingresses: { [i: string]: IResource };
+  deployments: { [d: string]: IKubeItem<IResource> };
```
Why do we need a map instead of an array? It seems that later we do `this.arrayFromState("services");`

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -207,26 +227,21 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
-                  Object.keys(this.state.ingresses).length > 0) && (
-                  <AccessURLTable services={this.state.services} ingresses={this.state.ingresses} />
-                )}
+                <AccessURLTable services={services} ingresses={ingresses} />
```
Much cleaner, thanks!

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -207,26 +227,21 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
-                />
+                <DeploymentsTable deployments={deployments} />
+                <ServicesTable services={services} />
+                <OtherResourcesTable otherResources={_.map(this.state.otherResources, r => r)} />
```
Why can't we use arrayFromState here?

migmartri commented on dashboard/src/components/AppView/DeploymentsTable/DeploymentsTable.tsx:
```diff
@@ -0,0 +1,46 @@
+          </thead>
+          <tbody>
+            {deployments.map(
+              d => d.item && <DeploymentItem key={d.item.metadata.name} deployment={d.item} />,
```
too late but I'd have called it `resource` instead of `item` :(

andresmgot replied:
I'd avoid to keep adding changes to this PR, if you think that's needed please open a different issue.

migmartri commented on dashboard/src/components/AppView/DeploymentsTable/DeploymentsTable.tsx:
```diff
@@ -0,0 +1,46 @@
+      <React.Fragment>
+        <h6>Deployments</h6>
+        <LoadingWrapper loaded={!isSomeResourceLoading(deployments)} size="small">
+          {deploymentSection}
```
another potentially cleaner way of doing this is by making deploymentSection a function. In any case, this looks quite good!

migmartri commented on dashboard/src/components/AppView/SecretsTable/SecretsTable.test.tsx:
```diff
@@ -64,3 +91,23 @@ it("renders a table with several secrets", () => {
   ).toBe(secret2);
 });
+
+it("renders a secret table with a secret and an error", () => {
```
Either implement ,delete, or mark this test case as pending (using `xit`)

andresmgot replied:
oops, I missed this

migmartri commented on dashboard/src/components/DeploymentStatus/DeploymentStatus.tsx:
```diff
@@ -1,25 +1,33 @@
   }
 
+  private renderLoading() {
+    return <span className="DeploymentStatus">Loading...</span>;
```
having a function for this only seems an overkill

andresmgot replied:
I just followed the style of this file. This specific method may be an overkill but not doing it may be inconsistent. Anyway, I am fine changing this.

migmartri commented on dashboard/src/components/DeploymentStatus/DeploymentStatus.tsx:
```diff
@@ -40,8 +48,11 @@ class DeploymentStatus extends React.Component<IDeploymentStatusProps> {
-        const status: IDeploymentStatus = d.status;
-        return status.availableReplicas === status.replicas;
+        if (d.item) {
+          const status: IDeploymentStatus = d.item.status;
```
This is the first time I see the every method. Interesting. You might want to explain that `every` returns false at the moment that the callback returns falsey.

andresmgot replied:
isn't it quite self-explanatory?

migmartri replied:
Well, it was not for me, but that can be my fault :)

migmartri commented on dashboard/src/containers/AppViewContainer/AppViewContainer.tsx:
```diff
@@ -30,6 +31,13 @@ function mapDispatchToProps(dispatch: ThunkDispatch<IStoreState, null, Action>)
     deleteApp: (releaseName: string, ns: string, purge: boolean) =>
       dispatch(actions.apps.deleteApp(releaseName, ns, purge)),
     getApp: (releaseName: string, ns: string) => dispatch(actions.apps.getApp(releaseName, ns)),
+    getResource: (
```
I'd have preferred to have this method, or even a specific one (see below) in the SecretsTableContainer instead.

```
# With specific method I meant this one in the SecretsTableContainer
getSecret(namespace, name)
```

migmartri replied:
BTW, for some reason I can not find where this method is being used.

andresmgot replied:
I forgot to remove this. You are right, it's used as `getSecret` in the SecretTableContainer

andresmgot commented:
I have applied your feedback @migmartri @prydonius. Thanks for the detailed comments!

migmartri commented on dashboard/src/components/AppView/AppView.tsx:
```diff
@@ -100,43 +104,46 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
-        return acc;
-      }, {});
+      // TODO: skip list resource for now
+      .filter(d => kindsWithTable.indexOf(d.kind) < 0 && d.kind !== "List");
```
It might make sense to use the switch case to also populate the otherResources?

Since the switch case below now it is also used to populate the initial state, it might make sense to use it for otherResources?

andresmgot replied:
good point

andresmgot commented:
I am merging this now. Let me know if there is still something you want me to change and I will open a different PR.
